insights/parsers/smt.py

```diff
@@ -0,0 +1,117 @@
+
+Parsers included in this module are:
+
+CpuSMTActive - file ``/sys/devices/system/cpu/smt/active``
+----------------------------------------------------------
+CpuCoreOnline - files matching ``/sys/devices/system/cpu/cpu[0-9]*/online``
+---------------------------------------------------------------------------
+CpuSiblings - files matching ``/sys/devices/system/cpu/cpu[0-9]*/topology/thread_siblings_list``
+------------------------------------------------------------------------------------------------
+"""
+
+from insights.core import Parser
+from insights.core.plugins import parser
+from insights.parsers import SkipException
+from insights.specs import Specs
+import re
+
+
+@parser(Specs.cpu_smt_active)
+class CpuSMTActive(Parser):
+    """
+    Class for parsing ``/sys/devices/system/cpu/smt/active`` file.
+    Reports whether SMT is enabled and active.
+
+    Typical output of this command is::
+
+        1
+
+    Raises:
+        SkipException: When content is empty or cannot be parsed.
+
+    Examples:
+        >>> cpu_smt.on
+        True
+    """
+    def parse_content(self, content):
+        if not content:
+            raise SkipException("No content.")
+        self.on = bool(int(content[0]))
+
+
+@parser(Specs.cpu_cores)
+class CpuCoreOnline(Parser):
+    """
+    Class for parsing ``/sys/devices/system/cpu/cpu[0-9]*/online`` matching files.
+    Reports whether a CPU core is online.
+
+    Typical output of this command is::
+
+        1
+        1
+        1
+
+    Raises:
+        SkipException: When content is empty or cannot be parsed
+
+    Examples:
+        >>> cpu_core.core_id
+        0
+        >>> cpu_core.on
+        True
+    """
+    cpu_core_path = r'/sys/devices/system/cpu/cpu(\d+)/online'
+
+    def parse_content(self, content):
+        if not content:
+            raise SkipException("No content.")
+        self.on = bool(int(content[0]))
+        self.core_id = int(re.match(CpuCoreOnline.cpu_core_path, self.file_path).group(1))
+
+    def __repr__(self):
+        return "[Core {0}: {1}]".format(self.core_id, "Online" if self.on else "Offline")
+
+
+@parser(Specs.cpu_siblings)
+class CpuSiblings(Parser):
+    """
+    Class for parsing ``/sys/devices/system/cpu/cpu[0-9]*/topology/thread_siblings_list`` matching files.
+    Reports CPU core siblings.
+
+    Typical output of this command is::
+
+        0,2
+        1,3
+        0,2
+        1,3
+
+    Raises:
+        SkipException: When content is empty or cannot be parsed
+
+    Examples:
+        >>> cpu_siblings.core_id
+        0
+        >>> cpu_siblings.siblings
+        [0, 2]
+    """
+    cpu_siblings_path = r'/sys/devices/system/cpu/cpu(\d+)/topology/thread_siblings_list'
+
+    def parse_content(self, content):
+        if not content:
+            raise SkipException("No content.")
+
+        # The separator in the sibling list may be either in the format 0-1 or 0,2 depending on the CPU model
+        if "-" in content[0]:
```

Comment:
Does this work correctly? Does not 0-2 actually mean 0, 1, and 2?

Reply:
I'm not quite sure but I think it works that way. See https://access.redhat.com/solutions/rhel-smt.

Reply:
I think that we may be lucky that it works because `-` separator was always used in the article or the tests like `0-1` or `2-3`, not for example `0-3`. @kgrant-rh any knowledge here?

Reply:
Based on the following I think that `-` is really supposed to be a range.
```
$ cat /sys/devices/system/cpu/cpu*/topology/core_siblings_list 
0-3
0-3
0-3
0-3
$ cat /sys/devices/system/cpu/cpu*/topology/thread_siblings_list 
0,2
1,3
0,2
1,3
```

Reply:
Yes, I believe you're correct, `-` is inclusive range, `,` lists individual items. I can't recall where in the source that behavior is specified, but I'll see if I can find it.

Reply:
Just for the record, this behavior is handled the same way as other bitmaps, as specified in /linux/lib/bitmap.c